File: src/main/java/de/espend/idea/vuejs/index/ComponentUsageIndex.java
package de.espend.idea.vuejs.index;

import com.intellij.util.indexing.*;
import com.intellij.util.io.DataExternalizer;
import com.intellij.util.io.EnumeratorStringDescriptor;
import com.intellij.util.io.KeyDescriptor;
import de.espend.idea.vuejs.index.externalizer.StringListDataExternalizer;
import de.espend.idea.vuejs.index.utils.VueJsComponentIndexUtil;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.vuejs.lang.html.VueFile;

import java.util.HashMap;
import java.util.List;
import java.util.Map;

/**
 * @author Daniel Espendiller <daniel@espendiller.net>
 */
public class ComponentUsageIndex extends FileBasedIndexExtension<String, List<String>> {
    public static final ID<String, List<String>> KEY = ID.create("de.espend.idea.vuejs.index.ComponentUsageIndex");
    private final KeyDescriptor<String> myKeyDescriptor = new EnumeratorStringDescriptor();

    @Override
    public @NotNull ID<String, List<String>> getName() {
        return KEY;
    }

    @Override
    public @NotNull DataIndexer<String, List<String>, FileContent> getIndexer() {
        return new DataIndexer<>() {
            @Override
            public @NotNull Map<String, List<String>> map(@NotNull FileContent fileContent) {
                Map<String, List<String>> imports = new HashMap<>();

                if (fileContent.getPsiFile() instanceof VueFile vueFile) {
                    imports.putAll(VueJsComponentIndexUtil.getComponentsForIndex(vueFile));
                    imports.putAll(VueJsComponentIndexUtil.getSameDirectoryComponentsForIndex(vueFile));
                }

                return imports;
            }
        };
    }

    @Override
    public @NotNull KeyDescriptor<String> getKeyDescriptor() {
        return this.myKeyDescriptor;
    }

    @Override
    public @NotNull DataExternalizer<List<String>> getValueExternalizer() {
        return StringListDataExternalizer.INSTANCE;
    }

    @Override
    public int getVersion() {
        return 1;
    }

    @NotNull
    @Override
    public FileBasedIndex.InputFilter getInputFilter() {
        return virtualFile -> "vue".equalsIgnoreCase(virtualFile.getExtension());
    }

    @Override
    public boolean dependsOnFileContent() {
        return true;
    }
}

File: src/main/kotlin/de/espend/idea/vuejs/index/ComponentUsageIndex.kt
package de.espend.idea.vuejs.index

import com.intellij.util.indexing.DataIndexer
import com.intellij.util.indexing.FileBasedIndex
import com.intellij.util.indexing.FileBasedIndexExtension
import com.intellij.util.indexing.FileContent
import com.intellij.util.indexing.ID
import com.intellij.util.io.DataExternalizer
import com.intellij.util.io.EnumeratorStringDescriptor
import com.intellij.util.io.KeyDescriptor
import de.espend.idea.vuejs.index.externalizer.StringListDataExternalizer
import de.espend.idea.vuejs.index.utils.VueJsComponentIndexUtil
import org.jetbrains.vuejs.lang.html.VueFile

class ComponentUsageIndex : FileBasedIndexExtension<String, List<String>>() {
    companion object {
        @JvmField
        val KEY: ID<String, List<String>> = ID.create("de.espend.idea.vuejs.index.ComponentUsageIndex")
    }

    private val myKeyDescriptor: KeyDescriptor<String> = EnumeratorStringDescriptor()

    override fun getName(): ID<String, List<String>> = KEY

    override fun getIndexer(): DataIndexer<String, List<String>, FileContent> = DataIndexer { fileContent ->
        val imports = mutableMapOf<String, List<String>>()
        val psiFile = fileContent.psiFile
        if (psiFile is VueFile) {
            imports.putAll(VueJsComponentIndexUtil.getComponentsForIndex(psiFile))
            imports.putAll(VueJsComponentIndexUtil.getSameDirectoryComponentsForIndex(psiFile))
        }
        imports
    }

    override fun getKeyDescriptor(): KeyDescriptor<String> = myKeyDescriptor

    override fun getValueExternalizer(): DataExternalizer<List<String>> = StringListDataExternalizer.INSTANCE

    override fun getVersion(): Int = 1

    override fun getInputFilter(): FileBasedIndex.InputFilter = FileBasedIndex.InputFilter { virtualFile ->
        "vue".equals(virtualFile.extension, ignoreCase = true)
    }

    override fun dependsOnFileContent(): Boolean = true
}
